src/utils/matching.ts

```diff
@@ -0,0 +1,119 @@
+const regexpCache = new Map();
+
+interface MatchInterface {
+    regexp: RegExp;
+    negated: boolean;
+}
+
+function makeRegexp(pattern: string): MatchInterface {
+    if (regexpCache.has(pattern)) {
+        return regexpCache.get(pattern);
+    }
+
+    const negated = pattern[0] === '!';
+    if (negated) {
+        pattern = pattern.substr(1);
+    }
+
+    // Check if the pattern is regex
+    if (pattern[0] === '/') {
+        // Get the flag of the specified regex
+        const flag = pattern.substr(pattern.lastIndexOf('/') + 1);
+        // Remove the / indentifiers so Regexp can make valid regexp of it.
+        pattern = pattern.substr(1).substr(0, pattern.lastIndexOf('/') - 1);
+        const regexp = new RegExp(pattern, flag);
+        const regObject: MatchInterface = {
+            regexp,
+            negated,
+        };
+        return regObject;
+    }
+    // Magic replacement to ensure pattern is valid
+    pattern = pattern.replace(/[|\\{}()[\]^$+*?.]/g, '\\$&').replace(/-/g, '\\x2d').replace(/\\\*/g, '[\\s\\S]*');
+    // Create an "hard" regexp to ensure that it's the exact domain we are matching and not some subdomain.
+    const regexp = new RegExp(`^${pattern}(?![A-Za-z0-9.])`, 'i');
+    const regObject: MatchInterface = {
+        regexp,
+        negated,
+    };
+    // Make sure it's cached!
+    regexpCache.set(pattern, regObject);
+    return regObject;
+}
+
+/**
+ * Sanitazed the website so it ensures different input
+ * with the same meaning will have the same output here.
+ */
+function sanitazeInput(input: string) {
+    return (input.replace(/^\^/, '')
+        .replace(/\$$/, '')
+        .replace(/\?.*$/, '')
+        .replace(/^.*?\/{2,3}/, '')
```

Comment:
Are there any URLs with `///` (three slashes) in them?

Reply:
It's a copy paste of some regex somewhere which we've already used. Which most likely was then created by @alexanderby :) So probably no, but maybe yes.

Reply:
I found out why it allows `///`: apparently, local files have URL like `file:///C:/stuff`. May be, we should add a comment or a unit test because I feel like this case could be lost during refactoring.

Reply:
Yeah, that's good to have a comment about that.